src/ct_selftest/ct_selftest.cpp

```diff
@@ -0,0 +1,297 @@
+   Botan::CT::poison(&poisoned_byte, sizeof(poisoned_byte));
+
+   std::array<uint8_t, 16> output_bytes;
+   std::memset(output_bytes.data(), 0x42, sizeof(output_bytes));
+
+   // Before the introduction of CT::value_barrier, this did generate a
+   // conditional jump when compiled with clang using certain compiler
+   // optimizations. See the test case above for further details.
+   auto poisoned_mask = Botan::CT::Mask<uint8_t>::expand(poisoned_byte & 1);
+   for(size_t i = 0; i < sizeof(output_bytes); ++i) {
+      output_bytes[i] = poisoned_mask.select(output_bytes[i], 0);
+   }
+
+   Botan::CT::unpoison(output_bytes.data(), output_bytes.size());
+   std::cout << Botan::hex_encode(output_bytes) << std::endl;
+}
+
+struct Test {
+      bool expect_failure;
+      std::function<void()> test;
+};
+
+constexpr bool SHOULD_FAIL = true;
+constexpr bool SHOULD_SUCCEED = false;
+
+void print_help(std::string_view path) {
+   std::cerr << "Usage: valgrind [...] " << path << " [testname]" << std::endl;
+   std::cerr << "Usage: " << path << " [--list|--list-special|--help]" << std::endl;
+   std::cerr << "By design, this can only run one test at a time."
+             << "Some tests are expected to cause CT::poison warnings and crashes." << std::endl;
+}
+
+void list_tests(const std::map<std::string, Test>& tests) {
+   std::cout << "fail?\ttest name\n\n";
+   for(const auto& [name, test_info] : tests) {
+      std::cout << Botan::fmt("{}\t{}\n", test_info.expect_failure ? "true" : "false", name);
+   }
+}
+
+template <typename MapT>
+auto merge_maps(const MapT& a, const MapT& b) {
+   MapT result = a;
+   result.insert(b.begin(), b.end());
+   return result;
+}
+
+}  // namespace
+
+int main(int argc, char* argv[]) {
+   // Those tests should run as expected on any compiler and any set of
+   // optimization flags.
+   std::map<std::string, Test> generic_tests = {
+      {"poisoned_conditional_jump", {SHOULD_FAIL, test_conditional_jump_on_poisoned_data}},
+      {"poisoned_memory_lookup", {SHOULD_FAIL, test_array_access_on_poisoned_data}},
+      {"transitive_poisoned_conditional_jump", {SHOULD_FAIL, test_conditional_jump_on_transitively_poisoned_data}},
+      {"transitive_poisoned_memory_lookup", {SHOULD_FAIL, test_array_access_on_transitively_poisoned_data}},
+      {"unpoison_transitive_poisoned", {SHOULD_SUCCEED, test_unpoisen_transitively_poisoned_data}},
+      {"poisoned_and_cleared", {SHOULD_SUCCEED, test_poisoned_and_cleared_data}},
+      {"conditional_jump_on_unpoisoned_bit", {SHOULD_SUCCEED, test_conditional_branch_on_unpoisoned_bit}},
+      {"conditional_jump_on_poisoned_bit", {SHOULD_FAIL, test_conditional_branch_on_poisoned_bit}},
+      {"regression_test_clang_vs_ct_mask", {SHOULD_SUCCEED, regression_test_conditional_jump_in_ct_mask}},
+   };
+
+   // These tests require special conditions to run. E.g. certain compiler
+   // versions and optimization flags.
+   std::map<std::string, Test> special_case_tests = {
```

Comment:
I don’t understand why the two maps here

Reply:
`ct_selftest.py` uses `ct_selftest --list` to detect which tests are available and runs them one-by-one using valgrind (assuming it to fail or not). The "special case tests" don't fail on all compilers with all flags, and hence need special treatment.

Alternatively, we could add another flag (next to "should fail") to signal that.

Reply:
> Alternatively, we could add another flag (next to "should fail") to signal that.

That's what I did, along with a slight refactor in `ct_selftest.PY` to make adding new "special case tests" easier and fail the test if one forgets to do so.